service/worker_master.py

```diff
@@ -126,131 +171,132 @@ def clean_pending_children(self):
+                HealthCheckCoroutine(
+                    api_key=self.api_key,
+                    node_id=self.node_id
+                ).run(websocket)
+            )
 
-                            raise TimeoutError()
+            # keep on processing commands from the server while possible
+            while websocket.open:
+                # cleanup pending_tasks
+                pending_children = self.clean_pending_children(pending_children)
 
-                        logger.info("Stopping tasks")
+                # process the next request
+                command_handler = await asyncio.ensure_future(self.process_command(websocket))
+                if command_handler and isinstance(command_handler, ProcessWrapperCommand):
+                    pending_children.append(command_handler)
 
-                        # send SIGSTOP to children
-                        for task in self.pending_children:
-                            await task.stop_process()
+        except Kill:
+            # prevent orphans
+            await WorkerManager.stop_all_children(pending_children)
 
-                        # setup timeout alarm
-                        signal.signal(signal.SIGALRM, alarm_handler)
-                        signal.alarm(10)
+            raise
+        finally:
+            if health_check_coroutine:
+                logger.info("Stopping Health check")
+                health_check_coroutine.cancel()
 
-                        # check that all things stopped
-                        for task in self.pending_children:
-                            await task.wait()
-                            logger.debug("Killed task %s", task)
+    async def run(self):
 
-                        # reset alarm so it doesn't fail
-                        signal.alarm(0)
+        reconnect = False # whether to try reconnecting with the same id
 
-                        raise
+        # keep a connection to a websocket while we're alive
+        while True:
 
-                    except:
-                        # print debugging info
-                        traceback.print_exc()
-                        raise
+            try:
 
-                logger.debug("Websocket dead")
+                # connects to websocket on host
+                async with websockets.connect(self.service_host) as websocket:
+                    await self.register_and_process(websocket, reconnect)
 
             except Kill:
                 logger.info("Shutting down node.")
                 sys.exit(0)
 
-            except:
-                if self.health_check_coroutine:
-                    self.health_check_coroutine.cancel()
-                    self.health_check_coroutine = None
+            except WebsocketConnectionError:
+                # print stack trace
+                traceback.print_exc()
 
+                self.num_reconnect_tries += 1
                 logger.info("Error occured, sleeping before trying to reconnect")
                 time.sleep(10)
```

Comment:
Instead try adaptive reconnect timeouts, i.e. sleep for 2^n * T seconds, where n is the number of attempted reconnects